bootstore/src/schemes/v0/state_initial_member.rs

```diff
@@ -0,0 +1,411 @@
+
+        let mut output = common.rack_secret_state.on_share(
+            from,
+            request_id,
+            share,
+            rack_secret_expiry,
+        );
+
+        if let Some(Ok(ApiOutput::RackSecret(rack_secret))) =
+            &mut output.api_output
+        {
+            // We have the rack secret as part of `output`. We may also
+            // have learn request that we need to resolve now.
+            //
+            // Note that we only persist if there are learn requests which
+            // cause an update to persistent state by handing out a share to the requester.
+            let (persist, envelopes) =
+                self.resolve_learn_requests(&rack_secret);
+            output.envelopes.extend_from_slice(&envelopes);
+            output.persist = persist;
+        }
+        output
+    }
+
+    // We've just recomputed the rack secret. Now resolve any pending learn
+    // requests. Return whether we need to persist state and any envelopes to
+    // send to peers.
+    fn resolve_learn_requests(
+        &mut self,
+        rack_secret: &RackSecret,
+    ) -> (bool, Vec<Envelope>) {
+        let mut persist = false;
+        let mut envelopes =
+            Vec::with_capacity(self.pending_learn_requests.len());
+        match self.pkg.decrypt_shares(rack_secret) {
+            Ok(shares) => {
+                while let Some((to, metadata)) =
+                    self.pending_learn_requests.pop_first()
+                {
+                    let Output {
+                        persist: persist_once,
+                        envelopes: envelopes_once,
+                        ..
+                    } = send_share_response(
+                        to.clone(),
+                        metadata.request_id,
+                        &self.pkg,
+                        &mut self.distributed_shares,
+                        &shares,
+                    );
+
+                    if persist_once {
+                        persist = true;
+                    }
+                    envelopes.extend(envelopes_once);
+                }
+            }
+            Err(_) => {
+                envelopes = self.resolve_learn_requests_with_error(
+                    Error::FailedToDecryptShares,
+                );
+            }
+        }
+
+        (persist, envelopes)
+    }
+
+    /// An error prevented successful resolution of learn requests
+    pub fn resolve_learn_requests_with_error(
+        &mut self,
+        error: Error,
+    ) -> Vec<Envelope> {
+        let mut envelopes =
+            Vec::with_capacity(self.pending_learn_requests.len());
+        while let Some((to, metadata)) = self.pending_learn_requests.pop_first()
+        {
+            let msg = Response {
+                request_id: metadata.request_id,
+                type_: error.into(),
+            };
+            envelopes.push(Envelope { to, msg: msg.into() })
+        }
+        envelopes
```

Comment:
```suggestion
        std::mem::take(&mut self.pending_learn_requests)
            .into_iter()
            .map(|(to, metadata)| {
                let msg = Response {
                    request_id: metadata.request_id,
                    type_: error.into(),
                };
                Envelope { to, msg: msg.into() }
            })
            .collect()
```

Reply:
No longer relevant as of [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)

This is similar to another comment of yours that I'm leaving open for fixes elsewhere.